### DISCUSSION TITLE
Combo with custom icon?

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [antvis/G6#3435] [Q&A]
Hi @Yanyan-Wang, 
1. Is it currently possible to render a combo with a custom icon like in the reference example below?

2. If not ^, what type of work would it take to implement this feature?
  - 2a. what area of the codebase would I need to look at to help enable this combo feature of using Image/Icon?  
  - 2b. what files do you think may need to be modified for this combo feature?
  - 2c. is this something on the G6 roadmap?
      - If not, maybe this can be a place to discuss how to approach implementation 🙏
  - 2d. What level of difficulty or size would you rate this task for a new contributor?
      - I've just begun starting to contribute to some of the Antv libraries like Graphin & I'm still fairly new to the world of non-DOM based rendering (svg/canvas based rendering), hence the reason trying to assess the level of effort required for this

### New Combo features?
- allow user to specify an image icon for the combo
- when the icon representing the combo clicks the image transitions away and the real combo content is revealed
![CleanShot 2022-01-10 at 19 20 40](https://user-images.githubusercontent.com/6743796/148875745-3030527a-5309-42b1-bf64-f470f031893c.gif)

- allow user to add a halo / icon to the edges of the combo (top-left, top-right, bottom-left, bottom-right)

https://user-images.githubusercontent.com/6743796/148876265-8cda468d-0d67-4afc-857e-667673b8676e.mp4

https://user-images.githubusercontent.com/6743796/148875939-d1ca9d20-7fcd-42e5-bf37-3a0a9f85fcc6.mp4

![image](https://user-images.githubusercontent.com/6743796/148875546-3ae5049e-39f3-4fb7-b9a8-5783560d7fd1.png)

Reference article:
- [Blog Article 1](https://cambridge-intelligence.com/combos/)
- [Blog Article 2](https://cambridge-intelligence.com/graph-visualization-rectangular-combos/)

**Comment by Yanyan-Wang** [comment]
Thanks @cliffordfajardo !

Combo in G6 is not perfect yet. I'm working on 2 related features recently:

1. combo layout kinda like the first video in your description: concentric layout inside + force layout outside. 
And I'm extending it to allow the user to configure the inside and outside layout freely. Such as grid layout inside + fa2 layout outside ……

2. support position assignment for combo. 
Currently, the position of the combo is totally depends on the distribution of its children. I'm trying to open the way for assigning the position for the combos directly, and its children will follow it in the same time.

Polishing the combo mechanism is one of the target of G6 in this year, and all the requirements, ideas, and contributions are welcome!! I will finish the two features I mentioned above in a month. To avoid the code conflicts, I suggest contributors to do something related after that.

Before that, maybe getting familiar with core/Item/node, core/item/combo, core/element/combo, and the APIs related in core/graph/graph and core/controller/item will be helpful. 

Thanks again!

**Comment by Yanyan-Wang** [comment]
We published G6 v4.6.0-beta.1 today (2022.2.11):
- combo supports position assignment (graph.updateItem)
- comboCombined layout: allow user to config the inner layout and outer layout of the combo. The innerLayout is concentric and the outerLayout is gForce by default. 

here goes a demo with uncombo/combo interactions, and the layout keeps stable

![combo-layout-stable-desensitized](https://user-images.githubusercontent.com/29593318/153575711-31c8eaba-8c65-4af6-a563-4c4d4fab7a2e.gif)


It is still a beta version, the usage could be found here:

- assign position for the combo, and the children follow it: https://github.com/antvis/G6/blob/19fa2d62924bc221eceb502b824348a42b095690/packages/core/tests/unit/shape/combo-collapsed-pos-spec.ts#L86

- comboForce Layout and part of interactions in the gif above: https://github.com/antvis/G6/blob/af2e3299019bbb247e967518262112359919feea/packages/pc/tests/unit/layout/combo-combined-spec.ts#L59

**Comment by cliffordfajardo** [comment]
Hi @Yanyan-Wang 
- Do of you have an example of creating a custom combo (`G6.registerCombo(...`) that will display a custom icon for the combo node?

@pomelo-nwu 
- do you happen to have an example of this in Graphin?

I'm having trouble finding an example online or when searching github for this use case custom icon for the combo (like in concept image shown above)
I started off with this starter code: 
- https://codesandbox.io/s/goofy-mountain-zhsh1u?file=/shape/combo/iconFont.js:1087-1109

@pomelo-nwu - mentioned this:
```
You can customize combo, add one more textShape, and use iconfont
```

**Comment by sklinov** [comment]
@cliffordfajardo We had a similar problem on our project but decided to migrate from combos to nodes with all the custom rendering we need. But that requires manual positioning of inner nodes (in our case it's good).
So now all items on the graph are first-class nodes, we are just passing some custom props to their configs for extra functionality.

**Reply by cliffordfajardo** [reply]
Hi @sklinov ,
Are you saying that you are emulating combo with icon behavior using a custom node you created?

```ts
G6.registerNode('node_name_', {
 //...custom code for node to emulate combo?
})
```

Do you mind sharing a video/image or code snippets (codesandbox.io?) of this? 
_Github supports dragging and dropping videos/images up to 10MB_ 

I'd like to adapt an example and to use as inspiration 🙏 


Bummer 😢 , combo's seem great, I wish you could just pass an `img` property to them similar to a node.

```
// Node: https://g6.antv.vision/en/examples/item/defaultNodes#image
{
      id: 'my_node1',
      img: 'https://gw.alipayobjects.com/os/s/prod/antv/assets/image/logo-with-text-73b8a.svg',
      x: 250,
      y: 150,
},
```

I wish combos allowed adding images this easily:
```
{
      id: 'my_combo1',
      img: 'https://gw.alipayobjects.com/os/s/prod/antv/assets/image/logo-with-text-73b8a.svg',
      // OR
      fontIcon: {
         text: 'fa-sitemap' // https://fontawesome.com/v4/icon/sitemap
         color: '#.....',
      }
      x: 250,
      y: 150,
},
```


> So now all items on the graph are first-class nodes, we are just passing some custom props to their configs for extra functionality.
Can you share more about this idea? This sounds interesting

Now that you are using nodes for all items, is there anything you feel you are missing out on that combo's had?
Ideally, what would you have wanted from the Combo API that perhaps would have saved you time from needing to create a custom node?

**Reply by sklinov** [reply]
@cliffordfajardo
Here's what we have now in development.
There are 2 types of nodes - one for "tables", and another for "columns".
So we're manually handling the size and positioning of everything (basically, it's the same `dagre.js` under the hood).
Sharing some code quickly might be complicated because it's a part of a large project.

https://user-images.githubusercontent.com/38185977/156525891-9268c098-70cb-440e-80e7-77d85c9c1c0f.mov

**Reply by sklinov** [reply]
@cliffordfajardo 
> Now that you are using nodes for all items, is there anything you feel you are missing out on that combo's had?

Not really, `combo`'s auto-resizing and positioning were PITA. 
For our requirements, I wish combos were more like nested nodes, but that's complicated and looks hard to be abstracted in a good way. 
I'm not a huge expert, but If you want to chat more about custom nodes feel free to reach out.

**Reply by cliffordfajardo** [reply]
Small progress on rendering custom Icons with Combos

G6 team mentioned this:
```
For custom for combos, similar to `G6.registerNode` and `G6.registerEdge`, you could  use the`addShape('image', ..)` method. in drawShape
```

Rendering custom Icon 🎉 - https://codesandbox.io/s/custom-combo-icon-image-sxwoql?file=/shape/combo/combo1.js:1305-1306

<img width="1291" alt="CleanShot 2022-03-07 at 18 56 53@2x" src="https://user-images.githubusercontent.com/6743796/157157449-b8f87c5e-bc66-43bd-90b3-a399a43f0df3.png">

.....I am just starting to learn the API for custom combo/edges/nodes in general & working towards the goal in the concept images above in the thread

**Reply by cliffordfajardo** [reply]
> I'm not a huge expert, but If you want to chat more about custom nodes feel free to reach out.

Hey @sklinov - I'd love to talk more about custom nodes, particularly nested nodes, I have not used that pattern yet. Let me know what works best for you.
I can hope on to a zoom video call, use a different chat app whatever works for you; I'll work on your time zone. My email is my github username at gmail dot com

Cheers 🎉

**Comment by Yanyan-Wang** [comment]
Hi, all. The combo supports image icon for collapsed status now (v4.6.8). See docs here: https://g6.antv.vision/en/docs/manual/middle/elements/combos/defaultCombo#collapsedsubstituteicon

**Comment by cliffordfajardo** [comment]
@Yanyan-Wang 

1. for the new `collapsedSubstitutionIcon` is there a way to a way in G6 to add  label indicators on the combo, for example this image below, without needing to use the `G6.registerNode`?

![CleanShot 2022-07-23 at 17 56 29@2x](https://user-images.githubusercontent.com/6743796/180627731-c3cfc46c-6549-4d70-923e-33ed58907a16.png)


2. Graphin has the concept of `badges` and `halos` do such concepts exist in G6 for small circular label that look like notification indicators on Combos & Node types?  

Code Sandbox Link: https://codesandbox.io/s/amazing-shape-fdvf5f?file=/App.tsx
![CleanShot 2022-07-23 at 18 04 02@2x](https://user-images.githubusercontent.com/6743796/180627892-4048fd4d-f072-4219-ab04-66bcd87c0281.png)

**Comment by Kogl1n** [comment]
Hi everyone! @Yanyan-Wang Did combos make it into Graphin? Would you please share a minimum example? Thank you for your effort!
